File: lib/data/models/candle.dart
import 'dart:math' as math;

class Candle {
  const Candle({
    required this.openTime,
    required this.open,
    required this.high,
    required this.low,
    required this.close,
    required this.volume,
    required this.closeTime,
    required this.quoteVolume,
    required this.takerBuyBaseVolume,
  });

  final int openTime;
  final double open;
  final double high;
  final double low;
  final double close;
  final double volume;
  final int closeTime;
  final double quoteVolume;
  final double takerBuyBaseVolume;

  double get typicalPrice => (high + low + close) / 3.0;
  double get range => high - low;
  bool get isBull => close >= open;
  double get body => (close - open).abs();
  double get upperWick => high - math.max(open, close);
  double get lowerWick => math.min(open, close) - low;

  /// Binance returns each kline as a JSON array:
  /// [openTime, open, high, low, close, volume, closeTime, quoteVolume, trades,
  ///  takerBuyBase, takerBuyQuote, ignore]
  static Candle fromArray(List<dynamic> a) {
    double d(dynamic v) => v is num ? v.toDouble() : double.parse(v.toString());
    int i(dynamic v) => v is num ? v.toInt() : int.parse(v.toString());
    return Candle(
      openTime: i(a[0]),
      open: d(a[1]),
      high: d(a[2]),
      low: d(a[3]),
      close: d(a[4]),
      volume: d(a[5]),
      closeTime: i(a[6]),
      quoteVolume: d(a[7]),
      takerBuyBaseVolume: d(a[9]),
    );
  }
}

File: app/src/main/kotlin/com/apex/trader/data/model/Candle.kt
package com.apex.trader.data.model

import com.apex.trader.data.api.dto.KlineDto

data class Candle(
    val openTime: Long,
    val open: Double,
    val high: Double,
    val low: Double,
    val close: Double,
    val volume: Double,
    val closeTime: Long,
    val quoteVolume: Double,
    val takerBuyBaseVolume: Double
) {
    val typicalPrice: Double get() = (high + low + close) / 3.0
    val range: Double get() = high - low
    val isBull: Boolean get() = close >= open
    val body: Double get() = kotlin.math.abs(close - open)
    val upperWick: Double get() = high - kotlin.math.max(open, close)
    val lowerWick: Double get() = kotlin.math.min(open, close) - low

    companion object {
        fun fromDto(d: KlineDto) = Candle(
            openTime = d.openTime,
            open = d.open,
            high = d.high,
            low = d.low,
            close = d.close,
            volume = d.volume,
            closeTime = d.closeTime,
            quoteVolume = d.quoteVolume,
            takerBuyBaseVolume = d.takerBuyBaseVolume
        )
    }
}

enum class Timeframe(val code: String, val millis: Long) {
    M1("1m", 60_000L),
    M5("5m", 300_000L),
    M15("15m", 900_000L),
    M30("30m", 1_800_000L),
    H1("1h", 3_600_000L),
    H4("4h", 14_400_000L),
    D1("1d", 86_400_000L);

    companion object {
        fun fromCode(code: String): Timeframe = entries.first { it.code == code }
    }
}
